tools/azure-sdk-tools/ci_tools/dependency_analysis.py

```diff
@@ -0,0 +1,350 @@
+
+                    packages[lib_name]["deps"].append({"name": req_name, "version": str(spec)})
+                    record_dep(dependencies, req_name, str(spec), lib_name)
+        except:
+            print("Failed to parse METADATA from %s" % (whl_path))
+    return packages, dependencies
+
+
+def dict_compare(d1: Dict[Any, Any], d2: Dict[Any, Any]) -> Tuple[Set[Any], Set[Any], Dict[Any, Tuple]]:
+    d1_keys = set(d1.keys())
+    d2_keys = set(d2.keys())
+    intersect_keys = d1_keys.intersection(d2_keys)
+    added = d1_keys - d2_keys
+    removed = d2_keys - d1_keys
+    modified = {o: (d1[o], d2[o]) for o in intersect_keys if d1[o] != d2[o]}
+    return added, removed, modified
+
+
+def render_report(output_path: str, report_context: Dict[str, any]) -> None:
+    env = Environment(loader=FileSystemLoader(os.path.dirname(os.path.realpath(__file__))))
+    template = env.get_template("deps.html.j2")
+    with io.open(output_path, "w", encoding="utf-8") as output:
+        output.write(template.render(report_context))
+
+
+def get_dependent_packages(data_pkgs) -> Set[str]:
+    # Get unique set of Azure SDK packages that are added as required package
+    deps = []
+    for v in data_pkgs.values():
+        deps.extend([dep["name"] for dep in v["deps"] if not dump_should_skip_lib(dep["name"])])
+    return set(deps)
+
+
+def dump_packages(data_pkgs: Dict[str, Dict[str, Any]]) -> Dict[str, Dict[str, Any]]:
+    dump_data = {}
+    unique_dependent_packages = get_dependent_packages(data_pkgs)
+    for p_name, p_data in data_pkgs.items():
+        p_id = p_name + ":" + p_data["version"]
+        dep = [p for p in p_data["deps"] if not dump_should_skip_lib(p["name"])]
+        # Add package if it requires other azure sdk package or if it is added as required by other sdk package
+        if len(dep) > 0 or p_name in unique_dependent_packages:
+            dump_data[p_id] = {"name": p_name, "version": p_data["version"], "type": "internal", "deps": dep}
+    return dump_data
+
+
+def resolve_lib_deps(dump_data: Dict[str, Dict[str, Any]], data_pkgs: Dict[str, Dict[str, Any]], pkg_id: str) -> None:
+    for dep in dump_data[pkg_id]["deps"]:
+        dep_req = Requirement.parse(dep["name"] + dep["version"])
+        if dep["name"] in data_pkgs and data_pkgs[dep["name"]]["version"] in dep_req:
+            # If the internal package version matches the dependency spec,
+            # rewrite the dep version to match the internal package version
+            dep["version"] = data_pkgs[dep["name"]]["version"]
+        else:
+            dep_id = dep["name"] + ":" + dep["version"]
+            if not dep_id in dump_data:
+                dump_data[dep_id] = {
+                    "name": dep["name"],
+                    "version": dep["version"],
+                    "type": "internalbinary" if dep["name"] in data_pkgs else "external",
+                    "deps": [],
+                }
+
+
+def pluralize(data: any, singular: str, plural: str) -> str:
```

Comment:
data: Any ?

Reply:
Something that works with `len`. If you know, I'll listen! Otherwise I'll spend some time digging for  this.

Reply:
Looks like I'm looking for [`collection.Sized` ](https://stackoverflow.com/a/49696395)